### ISSUE TITLE
[Bug Report] Calling SetReceiveMessageHandlerAsync(null, null) with Amqp will put connection in bad state

### ISSUE THREAD

**Issue by jamdavi** [Azure/azure-iot-sdk-csharp#2218]
# Context

- **OS, version, SKU and CPU architecture used:** Windows 10 Desktop x64
- **Application's .NET Target Framework :** netcore3.1
- **Device:** Laptop
- **SDK version used:** 1.38.0

## Description of the issue
When testing a detach operation by setting the handler to null the device client (and module client) will correctly detach but it will cause the client to disconnect and reconnect.

This is due to the device will call [SafeClose()](https://github.com/Azure/azure-iot-sdk-csharp/blob/5349c3f7b849c6ea65ac521510e5838c63fb532b/iothub/device/src/Transport/Amqp/AmqpUnit.cs#L230-L233) against the session when the receiver link is closed. It does this regardless of the reason it's called.

`Status Disconnected_Retrying, Reason Communication_Error`

## Code sample exhibiting the issue

```csharp
 static async Task Main(string[] args)
 {
        var devClient = DeviceClient.CreateFromConnectionString(<<CONNSTRING>>, TransportType.Amqp);
        await devClient.SetReceiveMessageHandlerAsync(MessageHandlerMethod, null);
        await Task.Delay(10000);
        await devClient.SetReceiveMessageHandlerAsync(null, null);
        await Task.Delay(int.MaxValue);
}
```

## Console log of the issue
```
2021-10-22T16:22:01.6224864>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (InternalClient#36038289, SetReceiveMessageHandlerAsync, ((null), (null))).
2021-10-22T16:22:01.6249817>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (RetryDelegatingHandler#49972132, DisableReceiveMessageAsync, (CancellationToken#3841804)).
2021-10-22T16:22:01.6264194>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (ErrorDelegatingHandler#65204782, ExecuteWithErrorHandlingAsync, ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync).
2021-10-22T16:22:01.6275157>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (AmqpTransportHandler#42750725, DisableReceiveMessageAsync, (CancellationToken#3841804)).
2021-10-22T16:22:01.6285700>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (AmqpUnit#39564799, DisableReceiveMessageAsync, (00:01:00)).
2021-10-22T16:22:01.6296226>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (AmqpUnit#39564799, DisableMessageReceivingLinkAsync, (00:01:00)).
2021-10-22T16:22:01.6304816>> Trace - [Microsoft-Azure-Amqp-AmqpLogOperationVerbose] (receiver7, 11, BeginClose).
2021-10-22T16:22:01.6423671>> Trace - [Microsoft-Azure-Amqp-AmqpStateTransition] (receiver7, S:DETACH, Opened, CloseSent).
2021-10-22T16:22:01.6437976>> Debug - [46EC.005 22:22:01.643] SEND FRM(0012|0000 detach(handle:1,closed:True))
2021-10-22T16:22:01.6449296>> Trace - [Microsoft-Azure-Amqp-AmqpLinkDetach] (receiver7, TelemetryReceiverLink, 1, S:DETACH, ).
2021-10-22T16:22:01.7059318>> Debug - [46EC.008 22:22:01.705] RECV FRM(0012|0000 detach(handle:1,closed:True))
2021-10-22T16:22:01.7072991>> Trace - [Microsoft-Azure-Amqp-AmqpLinkDetach] (receiver7, TelemetryReceiverLink, 1, R:DETACH, ).
2021-10-22T16:22:01.7081837>> Trace - [Microsoft-Azure-Amqp-AmqpStateTransition] (receiver7, R:DETACH, CloseSent, End).
2021-10-22T16:22:01.7096562>> Trace - [Microsoft-Azure-Amqp-AmqpRemoveLink] (out-connection0, session1, receiver7, 1, 1, TelemetryReceiverLink).
2021-10-22T16:22:01.7108357>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (AmqpIotReceivingLink#10168146, ReceivingAmqpLinkClosed, (ReceivingAmqpLinkClosed)).
2021-10-22T16:22:01.7123118>> Trace - [Microsoft-Azure-Amqp-AmqpLogOperationVerbose] (session1, 11, BeginClose).
2021-10-22T16:22:01.7128508>> Trace - [Microsoft-Azure-Amqp-AmqpLogOperationVerbose] (sender6, 11, BeginClose).
2021-10-22T16:22:01.7132629>> Trace - [Microsoft-Azure-Amqp-AmqpStateTransition] (sender6, S:DETACH, Opened, CloseSent).
2021-10-22T16:22:01.7136338>> Debug - [46EC.008 22:22:01.713] SEND FRM(0011|0000 detach(handle:0,closed:True))
2021-10-22T16:22:01.7138902>> Trace - [Microsoft-Azure-Amqp-AmqpLinkDetach] (sender6, TelemetrySenderLink, 0, S:DETACH, ).
2021-10-22T16:22:01.7147867>> Trace - [Microsoft-Azure-Amqp-AmqpStateTransition] (session1, S:END, Opened, CloseSent).
2021-10-22T16:22:01.7157493>> Debug - [46EC.008 22:22:01.715] SEND FRM(000F|0000 end())
2021-10-22T16:22:01.7161874>> Trace - [Microsoft-Azure-Devices-Device-Client-Exit] (AmqpIotReceivingLink#10168146, ReceivingAmqpLinkClosed, ReceivingAmqpLinkClosed).
2021-10-22T16:22:01.7166717>> Trace - [Microsoft-Azure-Amqp-AmqpLogOperationVerbose] (receiver7, 11, EndClose).
2021-10-22T16:22:01.7171666>> Trace - [Microsoft-Azure-Devices-Device-Client-Exit] (AmqpUnit#39564799, DisableMessageReceivingLinkAsync, 00:01:00).
2021-10-22T16:22:01.7210115>> Trace - [Microsoft-Azure-Devices-Device-Client-Exit] (AmqpUnit#39564799, DisableReceiveMessageAsync, 00:01:00).
2021-10-22T16:22:01.7216495>> Trace - [Microsoft-Azure-Devices-Device-Client-Exit] (AmqpTransportHandler#42750725, DisableReceiveMessageAsync, CancellationToken#3841804).
2021-10-22T16:22:01.7222496>> Trace - [Microsoft-Azure-Devices-Device-Client-Exit] (ErrorDelegatingHandler#65204782, ExecuteWithErrorHandlingAsync, ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync).
2021-10-22T16:22:01.7229171>> Trace - [Microsoft-Azure-Devices-Device-Client-Exit] (RetryDelegatingHandler#49972132, DisableReceiveMessageAsync, CancellationToken#3841804).
2021-10-22T16:22:01.7233247>> Trace - [Microsoft-Azure-Devices-Device-Client-Exit] (InternalClient#36038289, SetReceiveMessageHandlerAsync, (null), (null)).
2021-10-22T16:22:01.7745259>> Debug - [46EC.008 22:22:01.774] RECV FRM(0011|0000 detach(handle:0,closed:True))
2021-10-22T16:22:01.7752905>> Trace - [Microsoft-Azure-Amqp-AmqpLinkDetach] (sender6, TelemetrySenderLink, 0, R:DETACH, ).
2021-10-22T16:22:01.7758010>> Trace - [Microsoft-Azure-Amqp-AmqpStateTransition] (sender6, R:DETACH, CloseSent, End).
2021-10-22T16:22:01.7763041>> Trace - [Microsoft-Azure-Amqp-AmqpRemoveLink] (out-connection0, session1, sender6, 0, 0, TelemetrySenderLink).
2021-10-22T16:22:01.7769061>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (AmqpIotSendingLink#53488633, SendingAmqpLinkClosed, (SendingAmqpLinkClosed)).
2021-10-22T16:22:01.7777404>> Trace - [Microsoft-Azure-Devices-Device-Client-Exit] (AmqpIotSendingLink#53488633, SendingAmqpLinkClosed, SendingAmqpLinkClosed).
2021-10-22T16:22:01.7781871>> Trace - [Microsoft-Azure-Amqp-AmqpLogOperationVerbose] (sender6, 11, EndClose).
2021-10-22T16:22:01.7830607>> Debug - [46EC.008 22:22:01.782] RECV FRM(000F|0000 end())
2021-10-22T16:22:01.7836599>> Trace - [Microsoft-Azure-Amqp-AmqpStateTransition] (session1, R:END, CloseSent, End).
2021-10-22T16:22:01.7844892>> Trace - [Microsoft-Azure-Amqp-AmqpRemoveSession] (out-connection0, session1, 0, 0).
2021-10-22T16:22:01.7849838>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (AmqpIotSession#9922406, AmqpSessionClosed, (AmqpSessionClosed)).
2021-10-22T16:22:01.7854747>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (AmqpUnit#39564799, OnSessionDisconnected, (AmqpIotSession#9922406)).
2021-10-22T16:22:01.7859410>> Trace - [Microsoft-Azure-Devices-Device-Client-Info] (AmqpAuthenticationRefresher#31584737, StopLoop, StopLoop).
2021-10-22T16:22:01.8566635>> Trace - [Microsoft-Azure-Devices-Device-Client-Info] (AmqpTransportHandler#42750725, OnTransportDisconnected, OnTransportDisconnected).
2021-10-22T16:22:01.8581773>> Trace - [Microsoft-Azure-Devices-Device-Client-Info] (ProtocolRoutingDelegatingHandler#7244975, WaitForTransportClosedAsync, Client disconnected.).
2021-10-22T16:22:01.8598762>> Trace - [Microsoft-Azure-Devices-Device-Client-Info] (RetryDelegatingHandler#49972132, HandleDisconnectAsync, Transport disconnected: unexpected.).
2021-10-22T16:22:01.8608743>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (RetryStrategyAdapter#40677086, RetryStrategyAdapter.ShouldRetry, (0, Microsoft.Azure.Devices.Client.Exceptions.IotHubCommunicationException: Exception of type 'Microsoft.Azure.Devices.Client.Exceptions.IotHubCommunicationException' was thrown.)).
2021-10-22T16:22:01.8620570>> Trace - [Microsoft-Azure-Devices-Device-Client-Exit] (RetryStrategyAdapter#40677086, ShouldRetry, RetryStrategyAdapter.ShouldRetry).
2021-10-22T16:22:01.8626190>> Trace - [Microsoft-Azure-Devices-Device-Client-Exit] (AmqpUnit#39564799, OnSessionDisconnected, AmqpIotSession#9922406).
2021-10-22T16:22:01.8632694>> Trace - [Microsoft-Azure-Devices-Device-Client-Exit] (AmqpIotSession#9922406, AmqpSessionClosed, AmqpSessionClosed).
2021-10-22T16:22:01.8636757>> Trace - [Microsoft-Azure-Amqp-AmqpLogOperationVerbose] (session1, 11, EndClose).
2021-10-22T16:22:01.9686391>> Trace - [Microsoft-Azure-Devices-Device-Client-Enter] (InternalClient#36038289, OnConnectionStatusChanged, (Disconnected_Retrying, Communication_Error)).
2021-10-22T16:22:01.9702582>> Debug - Status Disconnected_Retrying, Reason Communication_Error
```

**Comment by timtay-microsoft** [comment]
Yep, it does appear that the link closed event handler for these links is closing the owning session even when the link is deliberately being closed by the client. Removing these handlers when deliberately closing the links will fix this issue. 

#2448

**Comment by timtay-microsoft** [comment]
This issue's fix has been released as of version 1.41.1 so I'm closing this thread
